### PR TITLE
libbpf-tools/profile: Add support for PID-namespacing

### PR DESCRIPTION
**Pull request by ekyooo**
Support PID namespace translation as memleak.py does.

### CODE REVIEW

File: libbpf-tools/profile.c [hunk 1]
```diff
@@ -595,6 +631,10 @@ int main(int argc, char **argv)
 				env.perf_max_stack_depth * sizeof(unsigned long));
 	bpf_map__set_max_entries(obj->maps.stackmap, env.stack_storage_size);
 
+	err = set_pidns(obj);
+	if (err)
```
Comment (yonghong-song) [inline libbpf-tools/profile.c:635]: Let us do 'err && env.verbose' here. We want to print out the below information under verbose mode.

File: libbpf-tools/profile.c [hunk 2]
```diff
@@ -507,10 +508,30 @@ static int print_counts(int counts_map, int stack_map)
...
 static int set_pidns(const struct profile_bpf *obj)
 {
 	struct stat statbuf;
 
+	/*
+	 * Available on Linux version 5.7 and above, which includes
+	 * the helper function bpf_get_ns_current_pid_tgid().
+	 */
+	if (kernel_at_least(5, 7))
+		return -EPERM;
```
[the remark this answers is not in the document]
Comment (yonghong-song) [reply libbpf-tools/profile.c:533]: In bcc/tools, we have a few cases (including profile) checking kernel version for bpf_get_ns_current_pid_tgid() availability. But for libbpf-tools, I think we can do better.
See trace_helpers.c. There are a few probe_*() functions which tries to detect whether a particular feature (prog_type, map_type) is supported or not. I think we can have another probe function to check whether bpf_get_ns_current_pid_tgid() is supported or not.

File: libbpf-tools/profile.c [hunk 3]
```diff
@@ -595,6 +610,8 @@ int main(int argc, char **argv)
 				env.perf_max_stack_depth * sizeof(unsigned long));
 	bpf_map__set_max_entries(obj->maps.stackmap, env.stack_storage_size);
 
+	set_pidns(obj);
```
[the remark this answers is not in the document]
Comment (ekyooo) [reply libbpf-tools/profile.c]: fixed. thank you.

File: libbpf-tools/profile.c (hunk 1 above)
Comment (ekyooo) [reply libbpf-tools/profile.c:635]: it's modified. thank you.

File: libbpf-tools/profile.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (ekyooo) [reply libbpf-tools/profile.c:533]: Thank you for good feedback.
I have applied the opinion.

@eiffel-fl 
Could you please check if it's what you want?

Thank you.

File: libbpf-tools/profile.bpf.c [hunk 4]
```diff
@@ -47,12 +50,23 @@ struct {
...
 	u64 *valp;
 	static const u64 zero;
 	struct key_t key = {};
+	u64 id;
+	u32 pid;
+	u32 tid;
+	struct bpf_pidns_info ns = {};
+
+	if (use_pidns && !bpf_get_ns_current_pid_tgid(pidns_dev, pidns_ino, &ns,
+						      sizeof(struct bpf_pidns_info))) {
```
[the remark this answers is not in the document]
Comment (ekyooo) [reply libbpf-tools/profile.bpf.c]: It's modified. Thank you.

File: libbpf-tools/trace_helpers.c [hunk 5]
```diff
@@ -1248,6 +1248,28 @@ bool probe_ringbuf()
 	return true;
 }
 
+bool probe_bpf_ns_current_pid_tgid(void)
```
[the remark this answers is not in the document]
Comment (ekyooo) [reply libbpf-tools/trace_helpers.c:1251]: It's done. thank you.

File: libbpf-tools/trace_helpers.c [hunk 6]
```diff
@@ -1248,6 +1248,28 @@ bool probe_ringbuf()
...
+		{ .code = BPF_ALU64 | BPF_MOV | BPF_X, .dst_reg = 3, .src_reg = BPF_REG_10 },
+		{ .code = BPF_ALU64 | BPF_ADD | BPF_K, .dst_reg = 3, .imm = -8 },
+		{ .code = BPF_ALU64 | BPF_MOV | BPF_K, .dst_reg = 1, .imm = 0 },
+		{ .code = BPF_ALU64 | BPF_MOV | BPF_K, .dst_reg = 2, .imm = 0 },
+		{ .code = BPF_ALU64 | BPF_MOV | BPF_K, .dst_reg = 4, .imm = 8 },
+		{ .code = BPF_JMP | BPF_CALL, .imm = BPF_FUNC_get_ns_current_pid_tgid },
+		{ .code = BPF_JMP | BPF_EXIT },
+	};
+
+	insn_cnt = sizeof(insns) / sizeof(struct bpf_insn);
```
[the remark this answers is not in the document]
Comment (ekyooo) [reply libbpf-tools/trace_helpers.c]: it's modified. Thank you.

**Comment by ekyooo** [comment]
> LGTM! Thank you!

Thank you.
I wonder if the test went well in your environment as well.

File: libbpf-tools/profile.bpf.c [hunk 7]
```diff
@@ -47,12 +50,23 @@ struct {
 SEC("perf_event")
 int do_perf_event(struct bpf_perf_event_data *ctx)
 {
-	u64 id = bpf_get_current_pid_tgid();
-	u32 pid = id >> 32;
-	u32 tid = id;
 	u64 *valp;
 	static const u64 zero;
 	struct key_t key = {};
+	u64 id;
```
[the remark this answers is not in the document]
Comment (ekyooo) [reply libbpf-tools/profile.bpf.c:56]: Yes, I prefer that style, but as far as I know, bcc/libbpf-tools follows the 'declaration first' rule.
Thank you for good opinion.

**Pull request merged**